pykwasm/src/pykwasm/kdist/wasm-semantics/numeric.md

```diff
@@ -136,18 +136,18 @@ Note: The actual `ctz` operator considers the integer 0 to have *all* zero-bits,
     rule ITYPE . ctz    I1 => < ITYPE > #if I1 ==Int 0 #then #width(ITYPE) #else #ctz(I1) #fi
     rule ITYPE . popcnt I1 => < ITYPE > #popcnt(I1)
 
-    syntax Int ::= #minWidth ( Int ) [function]
-                 | #ctz      ( Int ) [function]
-                 | #popcnt   ( Int ) [function]
+    syntax Int ::= #minWidth ( Int ) [function, total]
+                 | #ctz      ( Int ) [function, total]
+                 | #popcnt   ( Int ) [function, total]
  // -------------------------------------------
-    rule #minWidth(0) => 0
-    rule #minWidth(N) => 1 +Int #minWidth(N >>Int 1)                                 requires N =/=Int 0
+    rule #minWidth(N) => 0                                                           requires N <=Int 0
+    rule #minWidth(N) => 1 +Int #minWidth(N >>Int 1)                                 requires N >Int 0
 
-    rule #ctz(0) => 0
-    rule #ctz(N) => #if N modInt 2 ==Int 1 #then 0 #else 1 +Int #ctz(N >>Int 1) #fi  requires N =/=Int 0
+    rule #ctz(N) => 0                                                                requires N <=Int 0
+    rule #ctz(N) => #if N modInt 2 ==Int 1 #then 0 #else 1 +Int #ctz(N >>Int 1) #fi  requires N >Int 0
 
-    rule #popcnt(0) => 0
-    rule #popcnt(N) => #bool(N modInt 2 ==Int 1) +Int #popcnt(N >>Int 1)             requires N =/=Int 0
+    rule #popcnt(N) => 0                                                             requires N <=Int 0
```

Comment:
@virgil-serbanuta @jberthold In the official wasm spec, these functions are defined only for non-neative numbers (undefined for negative numbers) because integer values are kept as uninterpreted integers. This means that the value in `<i32|i64> N` is always non-negative and converted to a signed int depending on the instruction's signedness. 

* [Values](https://webassembly.github.io/spec/core/exec/runtime.html#values)
* [Integer values](https://webassembly.github.io/spec/core/syntax/values.html#syntax-int)

You can check out the implementation of [`lt_s`](https://github.com/runtimeverification/wasm-semantics/blob/master/pykwasm/src/pykwasm/kdist/wasm-semantics/numeric.md?plain=1#L387) for how unsigned arguments are processed as signed integers using `#signed(_,_)`. 

Ideally I would prefer to use natural numbers instead of `Int` but we don't have built-in natural numbers.

For these reasons, I decided to return a default value of `0` for negative numbers in these functions to keep the implementation as simple as possible. Maybe to be more defensive I can add a check to the instruction's rule to reject negative numbers.

Reply:
Maybe the best way forward is indeed to leave these 3 functions the way they were proposed and protect each call site to ensure the argument is non-negative.

One could also make the 3 functions return `-1` for negative arguments, which makes it obvious that the result is invalid (but of course does not protect anything).

Reply:
List of updates after your reviews:
* `#minWidth`, `#ctz`, `#popcnt` returns `-1` for negative numbers to indicate a failure. I think this will be better than `0` when debugging if anything goes wrong.
* The `T:IValType . OP:IUnOp I1` function returns `undefined` for negative numbers.
* Replaced `modInt 2` with `&Int 1` for `&Int` is a total function.